get operational status and multicast support for osx

wfurt opened the pull request:
fixes #29723

stephentoub commented on src/Native/Unix/System.Native/pal_networkstatistics.c:
```diff
@@ -38,6 +38,12 @@
+enum {
+    InteropInterfaceUp = 0x1,
+    InteropSupportsMulticast = 0x2
+};
```
Shouldn't this be in pal_networkstatistics.h?  Seems like it's effectively part of TcpGlobalStatistics's definition.

stephentoub commented on src/Native/Unix/System.Native/pal_networkstatistics.c:
```diff
@@ -482,6 +488,18 @@ int32_t SystemNative_GetNativeIPInterfaceStatistics(char* interfaceName, NativeI
+                retStats->Flags |= InteropInterfaceUp;
+            }
+
+            if (ifHdr->ifm_flags & (IFF_MULTICAST | IFF_ALLMULTI))
```
This is currently stating that multicast is supported if either the IFF_MULTICAST or the IFF_ALLMULTI flag is set... it's correct that's it's either, rather than, say, requiring both?

wfurt replied:
correct. When IFF_MULTICAST is set you can receive traffic for subscribed multicast addresses. 
When IFF_ALLMULTI you can receive traffic even for groups you did not subscribed to. Either one is sign that multicast operations are supported. When neither flag is set, setting subscription would probably fail. (at least is did on Linux)

stephentoub commented on src/Native/Unix/System.Native/pal_networkstatistics.c:
```diff
@@ -38,6 +38,12 @@
+
+enum {
+    InteropInterfaceUp = 0x1,
+    InteropSupportsMulticast = 0x2
```
What's the meaning of "Interop" in the naming here?

stephentoub commented:
Is there a test we can add?

wfurt commented:
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

wfurt commented on src/Native/Unix/System.Native/pal_networkstatistics.c:
```diff
@@ -482,6 +488,18 @@ int32_t SystemNative_GetNativeIPInterfaceStatistics(char* interfaceName, NativeI
             retStats->InNoProto = systemStats.ifi_noproto;
+
+            retStats->Flags = 0;
+            if (ifHdr->ifm_flags & IFF_UP)
```
I don't think IFF_RUNNING guaranties link. I get this after unplugging cable:
```
en8: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=b<RXCSUM,TXCSUM,VLAN_HWTAGGING>
	ether 00:24:9b:07:be:11
	nd6 options=201<PERFORMNUD,DAD>
	media: autoselect (<unknown type>)
	status: inactive
```  

If we want to track link state I can extend it more. I think we would need to make another call to query link - at least for types which do support it.

wfurt replied:
it seems like that is what Linux does e.g. interface state tracks actual link. I'll make some updates to be consistent.

wfurt commented:
I added detection for media/link status. That part is OSX/BSD specific so I added new define to do conditional compilation. (This won't be present on Linux) 
Additionally I made change to pass interface index to sysctl so we should get only requested interface instead of getting full list every time.


With this change I get now:
````
lo0: Up Loopback True sp=0
gif0: Down Tunnel True sp=0
stf0: Down Tunnel False sp=0
XHC0: Down Unknown False sp=0
XHC1: Down Unknown False sp=0
XHC20: Down Unknown False sp=0
en0: Up Ethernet True sp=304200000
p2p0: Down Ethernet True sp=10000000
awdl0: Up Ethernet True sp=10000000
en3: Down Ethernet True sp=0
en4: Down Ethernet True sp=0
en1: Down Ethernet True sp=0
en2: Down Ethernet True sp=0
bridge0: Down Ethernet True sp=0
utun0: Up Unknown True sp=0
utun1: Up Unknown True sp=0
vnic0: Up Ethernet True sp=0
vnic1: Up Ethernet True sp=0
gpd0: Down Ethernet True sp=0
en5: Up Ethernet True sp=0
en8: Up Ethernet True sp=1000000000
```
and it changes to 
```
en8: Down Ethernet True sp=1000000000
```
when I pull cable on en8

wfurt commented:
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

stephentoub commented on src/Native/Unix/System.Native/pal_networkstatistics.c:
```diff
@@ -482,6 +486,54 @@ int32_t SystemNative_GetNativeIPInterfaceStatistics(char* interfaceName, NativeI
+
+                    if ((ioctl(fd, SIOCGIFMEDIA, (caddr_t)&ifmr) < 0) || ((ifmr.ifm_status & IFM_AVALID) == 0))
+                    {
+                        if (errno == EOPNOTSUPP)
```
Is it possible if ioctl succeeded (but the subsequent check failed) that errno could still be EOPNOTSUPP from some previous, unrelated operation?  Or does ioctl guarantee errno will be reset to 0 or something if it succeeds?

wfurt replied:
system call would set errno on failure but if it succeeds value of errno is not defined e.g. it could have old value as you suggested. I'll fix it.

stephentoub commented:
Do we already have tests that exercise this?  If not, can we add some?

wfurt commented:
the linked #29723 is about network availability change. That is hard to do as unit test.
I'll think about it and maybe I can add test to verify that known interfaces do not have unknown state. 
If you have some other suggestions, please let me know @stephentoub

stephentoub commented:
> That is hard to do as unit test.

Sure, but we should be able to at least enumerate all of this information in a unit test, right?  To at least ensure the code paths are exercised / don't crash / etc.?  We may not be able to verify that all of the data has the expected results, but we should be able to verify lack of failure and lack of out-of-range results.
